turbine/src/quic_endpoint.rs

```diff
@@ -0,0 +1,429 @@
+        error!("handle_connecting: {err:?}");
+    }
+}
+
+async fn handle_connecting(
+    connecting: Connecting,
+    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+    cache: Arc<RwLock<ConnectionCache>>,
+) -> Result<(), Error> {
+    let connection = connecting.await?;
+    let remote_address = connection.remote_address();
+    let remote_pubkey = get_remote_pubkey(&connection)?;
+    handle_connection_error(remote_address, remote_pubkey, connection, sender, cache).await;
+    Ok(())
+}
+
+async fn handle_connection_error(
+    remote_address: SocketAddr,
+    remote_pubkey: Pubkey,
+    connection: Connection,
+    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+    cache: Arc<RwLock<ConnectionCache>>,
+) {
+    cache_connection(remote_address, remote_pubkey, connection.clone(), &cache).await;
+    if let Err(err) = handle_connection(remote_address, remote_pubkey, &connection, sender).await {
+        drop_connection(remote_address, remote_pubkey, &connection, &cache).await;
+        error!("handle_connection: {remote_pubkey}, {remote_address}, {err:?}");
+    }
+}
+
+async fn handle_connection(
+    remote_address: SocketAddr,
+    remote_pubkey: Pubkey,
+    connection: &Connection,
+    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+) -> Result<(), Error> {
+    // Assert that send won't block.
+    debug_assert_eq!(sender.capacity(), None);
+    loop {
+        match connection.read_datagram().await {
+            Ok(bytes) => sender.send((remote_pubkey, remote_address, bytes))?,
+            Err(err) => {
+                if let Some(err) = connection.close_reason() {
+                    return Err(Error::from(err));
+                }
+                error!("connection.read_datagram: {remote_pubkey}, {remote_address}, {err:?}");
+            }
+        };
+    }
+}
+
+async fn send_datagram_task(
+    endpoint: Endpoint,
+    remote_address: SocketAddr,
+    bytes: Bytes,
+    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+    cache: Arc<RwLock<ConnectionCache>>,
+) {
+    if let Err(err) = send_datagram(&endpoint, remote_address, bytes, sender, cache).await {
+        error!("send_datagram: {remote_address}, {err:?}");
+    }
+}
+
+async fn send_datagram(
+    endpoint: &Endpoint,
+    remote_address: SocketAddr,
+    bytes: Bytes,
+    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+    cache: Arc<RwLock<ConnectionCache>>,
+) -> Result<(), Error> {
+    let connection = get_connection(endpoint, remote_address, sender, cache).await?;
+    connection.send_datagram(bytes)?;
+    Ok(())
+}
+
+async fn get_connection(
+    endpoint: &Endpoint,
+    remote_address: SocketAddr,
+    sender: Sender<(Pubkey, SocketAddr, Bytes)>,
+    cache: Arc<RwLock<ConnectionCache>>,
+) -> Result<Connection, Error> {
+    let key = (remote_address, /*remote_pubkey:*/ None);
+    let entry = cache.write().await.entry(key).or_default().clone();
+    if let Some(connection) = entry.read().await.deref() {
```

Comment:
Are async locks also potentially held for too long, like in https://github.com/solana-labs/solana/issues/24836 ? If so, might be worth scoping this to avoid potentially deadlocking

Reply:
done.